syncer/syncer.go

```diff
@@ -231,38 +247,158 @@ func (s *Syncer) processBlock(
+		}
+	}
+	return nil
+}
+
+// fetchChannelBlocks fetches blocks from a
+// channel with retries until there are no
+// more blocks in the channel or there is an
+// error.
+func (s *Syncer) fetchChannelBlocks(
+	ctx context.Context,
+	network *types.NetworkIdentifier,
+	blockIndicies chan int64,
+	results chan *types.Block,
+) error {
+	for b := range blockIndicies {
+		block, err := s.helper.Block(
+			ctx,
+			network,
+			&types.PartialBlockIdentifier{
+				Index: &b,
+			},
+		)
+		if err != nil {
+			return err
+		}
+
+		select {
+		case results <- block:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
 	}
 
+	return nil
+}
+
+// processBlocks is invoked whenever a new block is fetched. It attempts
+// to process as many blocks as possible.
+func (s *Syncer) processBlocks(
+	ctx context.Context,
+	cache map[int64]*types.Block,
+	endIndex int64,
+) error {
+	// We need to determine if we are in a reorg
+	// so that we can force blocks to be fetched
+	// if they don't exist in the cache.
+	reorgStart := int64(-1)
+
 	for s.nextIndex <= endIndex {
-		block, ok := blockMap[s.nextIndex]
-		if !ok { // could happen in a reorg
-			block, err = s.fetcher.BlockRetry(
+		block, exists := cache[s.nextIndex]
+		if !exists {
+			// Wait for more blocks if we aren't
+			// in a reorg.
+			if reorgStart < s.nextIndex {
+				break
+			}
+
+			// Fetch the nextIndex if we are
+			// in a re-org.
+			newBlock, err := s.helper.Block(
 				ctx,
 				s.network,
 				&types.PartialBlockIdentifier{
 					Index: &s.nextIndex,
 				},
 			)
 			if err != nil {
-				return err
+				return fmt.Errorf("%w: unable to fetch block %d", err, s.nextIndex)
 			}
+
+			block = newBlock
 		} else {
 			// Anytime we re-fetch an index, we
 			// will need to make another call to the node
 			// as it is likely in a reorg.
-			delete(blockMap, s.nextIndex)
+			delete(cache, s.nextIndex)
 		}
 
-		if err = s.processBlock(ctx, block); err != nil {
-			return err
+		lastProcessed := s.nextIndex
+		if err := s.processBlock(ctx, block); err != nil {
+			return fmt.Errorf("%w: unable to process block", err)
 		}
+
+		if s.nextIndex < lastProcessed && reorgStart == -1 {
+			reorgStart = lastProcessed
+		}
+	}
+
+	return nil
+}
+
+func (s *Syncer) syncRange(
+	ctx context.Context,
+	endIndex int64,
+) error {
+	blockIndicies := make(chan int64)
+	results := make(chan *types.Block)
+	g, ctx := errgroup.WithContext(ctx)
+	g.Go(func() error {
+		return s.addBlockIndicies(ctx, blockIndicies, s.nextIndex, endIndex)
+	})
+
+	for j := uint64(0); j < s.concurrency; j++ {
+		g.Go(func() error {
+			return s.fetchChannelBlocks(ctx, s.network, blockIndicies, results)
+		})
+	}
+
+	// Wait for all block fetching goroutines to exit
+	// before closing the results channel.
+	go func() {
+		_ = g.Wait()
+		close(results)
+	}()
+
+	cache := make(map[int64]*types.Block)
+	for b := range results {
+		cache[b.BlockIdentifier.Index] = b
+
+		if err := s.processBlocks(ctx, cache, endIndex); err != nil {
```

Comment:
just curious why do we need to iterate over the results while `s.processBlocks` is also processing from s.nextIndex to endIndex?

Reply:
We are pulling from an unordered channel here. Here is an example:
```
s.nextIndex = 7
result -> block 10
result -> block 8
result -> block 9
result -> block 7 
-> process block 7-10
```

Reply:
We want to process all retrieved but unprocessed blocks each time we take something out of the channel.

Reply:
Definitely the most complex part of this implementation (which is why I wrote so many tests).